src/utils/src/byte_order.rs

```diff
@@ -47,35 +47,119 @@ generate_write_fn!(write_be_u32, u32, u8, to_be_bytes);
 mod tests {
     use super::*;
     macro_rules! byte_order_test_read_write {
-        ($test_name: ident, $write_fn_name:ident, $read_fn_name:ident, $is_be: expr, $data_type: ty, $byte_type: expr, $type_size: expr ) => {
-            #[allow(overflowing_literals)]
+        ($test_name: ident, $write_fn_name: ident, $read_fn_name: ident, $validate_le: expr, $validate_be: expr, $data_type: ty, $byte_type: expr, $type_size: expr ) => {
             #[test]
             fn $test_name() {
-                let le_original = [0xef, 0xcd, 0xab, 0x89, 0x67, 0x45, 0x23, 0x01];
-                let be_original = [0x01, 0x23, 0x45, 0x67, 0x89, 0xab, 0xcd, 0xef];
-                let v = 0x0123_4567_89AB_CDEF as $data_type;
-                let cmp_arr = if $is_be {
-                    &be_original[(8 - $type_size)..]
-                } else {
-                    &le_original[..]
-                };
-                // test write
-                let mut writearr = [$byte_type; $type_size];
-                $write_fn_name(&mut writearr, v);
-                for (cmp, cur) in cmp_arr.iter().zip(writearr.iter().cloned()) {
-                    assert_eq!(*cmp, cur as u8)
+                #[allow(overflowing_literals)]
+                let test_cases = [
+                    (
+                        0x0123_4567_89AB_CDEF as u64,
+                        [0x01, 0x23, 0x45, 0x67, 0x89, 0xab, 0xcd, 0xef],
+                    ),
+                    (
+                        0x0000_0000_0000_0000 as u64,
+                        [0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00],
+                    ),
+                    (
+                        0x1923_2345_ABF3_CCD4 as u64,
+                        [0x19, 0x23, 0x23, 0x45, 0xAB, 0xF3, 0xCC, 0xD4],
+                    ),
+                    (
+                        0x0FF0_0FF0_0FF0_0FF0 as u64,
+                        [0x0F, 0xF0, 0x0F, 0xF0, 0x0F, 0xF0, 0x0F, 0xF0],
+                    ),
+                    (
+                        0xFFFF_FFFF_FFFF_FFFF as u64,
+                        [0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF],
+                    ),
+                    (
+                        0x89AB_12D4_C2D2_09BB as u64,
+                        [0x89, 0xAB, 0x12, 0xD4, 0xC2, 0xD2, 0x09, 0xBB],
+                    ),
+                ];
+
+                for (test_val, v_arr) in &test_cases {
+                    let v = *test_val as $data_type;
+                    let cmp_iter: Box<dyn Iterator<Item = _>> = if $validate_le {
+                        Box::new(v_arr[..].iter().rev())
+                    } else if $validate_be {
+                        Box::new(v_arr[(8 - $type_size)..].iter())
+                    } else {
+                        assert!(false, "test must either validate le or be");
+                        Box::new(v_arr[..].iter())
+                    };
```

Comment:
I think it would make more sense for `validate_le` and `validate_be` not to be mutually exclusive. You should be able to validate le or be or both.

Reply:
Yes, I agree but I am not quite sure how to implement it. I sent this message above:

The issue I ran into was using the same macro invocation to test both big-endian and little-endian since each requires a unique function call (write_be_u32 vs write_le_u32), meaning that testing both versions in the same macro call requires both versions of the function call to be passed to the macro. Is there a way to construct a function call inside the macro given only the data type? For example, is there a way to pass in u32 as a parameter to the macro and make a function call to write_le_u32 and write_be_u32?

Reply:
Ok, I see. I missed this before. I will think about it.

Reply:
I can't think of a solution to this. From what I know you can't obtain the function name by concatenating strings. You have to provide it explicitly in the macro call. Then probably it's better to return to the previous version where you had only one flag (`is_be`).